src/util/logging/logger.go

```diff
@@ -1,172 +1,71 @@
-		Level:     logrus.FatalLevel,
-	}
-)
+// NewMasterLogger creates a new package-aware logger with formatting string
+func NewMasterLogger() *MasterLogger {
+	hooks := make(logrus.LevelHooks)
 
-// NewLogger creates a new modules-aware logger with formatting string
-func NewLogger(priorityKey, criticalPriority string) (logger *Logger) {
-	logger = &Logger{
+	return &MasterLogger{
 		Logger: &logrus.Logger{
 			Out: os.Stdout,
 			Formatter: &TextFormatter{
-				FullTimestamp:          true,
-				AlwaysQuoteStrings:     true,
-				QuoteEmptyFields:       true,
-				ForceFormatting:        true,
-				PriorityKey:            priorityKey,
-				HighlightPriorityValue: criticalPriority,
+				FullTimestamp:      true,
+				AlwaysQuoteStrings: true,
+				QuoteEmptyFields:   true,
+				ForceFormatting:    true,
+				DisableColors:      false,
+				ForceColors:        false,
 			},
-			Hooks: make(logrus.LevelHooks),
-			Level: logrus.InfoLevel,
-		},
-		allModulesEnabled: true,
-		moduleLoggers:     make(map[string]*Logger),
-		PriorityKey:       priorityKey,
-		CriticalPriority:  criticalPriority,
-	}
-	logger.Hooks.Add(NewModuleLogHook(""))
-	logger.moduleLoggers[""] = logger
-	return
-}
-
-// LoggerForModules creates a logger for a set of modules
-func LoggerForModules(priorityKey, criticalPriority string, enabledModules []string) *Logger {
-	logger := NewLogger(priorityKey, criticalPriority)
-	logger.allModulesEnabled = false
-	for _, moduleName := range enabledModules {
-		// Lazy instantiation
-		logger.moduleLoggers[moduleName] = nil
-	}
-	return logger
-}
-
-func (logger *Logger) cloneForModule(moduleName string) *Logger {
-	newLogger := &Logger{
-		Logger: &logrus.Logger{
-			Out:       logger.Out,
-			Formatter: logger.Formatter,
-			Hooks:     make(logrus.LevelHooks),
-			Level:     logger.Level,
+			Hooks: hooks,
+			Level: logrus.DebugLevel,
 		},
-		allModulesEnabled: logger.allModulesEnabled,
-		moduleLoggers:     logger.moduleLoggers,
-		PriorityKey:       logger.PriorityKey,
-		CriticalPriority:  logger.CriticalPriority,
 	}
-	newLogger.Hooks.Add(NewReplayHook(logger.moduleLoggers[""]))
-	newLogger.Hooks.Add(NewModuleLogHook(moduleName))
-	return newLogger
 }
 
-// MustGetLogger returns an existing logger for a given module or creates a new one
-func (logger *Logger) MustGetLogger(moduleName string) *Logger {
-	newLogger, isInCache := logger.moduleLoggers[moduleName]
-	if !(isInCache && newLogger != nil) {
-		if isInCache || logger.allModulesEnabled {
-			newLogger = logger.cloneForModule(moduleName)
-			logger.moduleLoggers[moduleName] = newLogger
-		} else {
-			newLogger = &Logger{
-				Logger:            &QuietLogger,
-				allModulesEnabled: logger.allModulesEnabled,
-				moduleLoggers:     logger.moduleLoggers,
-			}
-		}
+// PackageLogger instantiates a package-aware logger
+func (logger *MasterLogger) PackageLogger(moduleName string) *Logger {
+	return &Logger{
+		FieldLogger: logger.WithField(logModuleKey, moduleName),
```

Comment:
This is something I considered doing when submitting the patch for #858 . I noticed that there is a difference between logger and entry instances , even if they both implement `FieldLogger` interface. Instances of the former are meant to be stand-alone objects whereas the later are stored in a pool and eventually released back to the pool.

Since [`WithField` method returns an entry](https://github.com/sirupsen/logrus/blob/master/logger.go#L93-L105) , it was not clear to me whether this object interacted with the pool somehow . If that was the case then it might be put back into the logger pool . It was not clear to me either whether reused entries could be polluted from one call to the next one . Another scenario that came into my mind was a change applied upon the `Data` of the persistent entry referenced by our `PackaLogger` instances. This would be definitely feasible and would change significantly logger behavior , thus leading to obscure errors very hard to detect .

That's why I didn't follow that path and stick to logger instances instead of keeping a reference to entries beyond the life-time of the `FieldLogger` method calls. In the previous implementation entries were ephemeral , and I'd recommend to keep it that way so as to avoid headaches . Nonetheless , if this is no

Reply:
It turns out it is unsafe to modify entry.Data ourselves as was done in the ModuleHook (it paniced on a concurrent map write once I condensed everything into MasterLogger, I checked previous github issues and the author mentioned Hooks are not meant to write to Data), hence the prefix handling is moved to the formatter.

The entry pooling shouldn't have any effect on this

Reply:
👍